### ISSUE TITLE
Unable to load dynamic imports with vite

### ISSUE THREAD

**Issue by DinosaurDad** [ionic-team/pwa-elements#109]
**Resources:**

```
dependencies:
@capacitor/camera 1.3.1
@capacitor/core 3.4.3
@ionic/pwa-elements 3.1.1
@ionic/vue 6.0.13
vue 3.2.31
vue-router 4.0.14

devDependencies:
@vitejs/plugin-vue 2.2.4
vite 2.8.6
```

**I'm submitting a ...**  (check one with "x")
[x] bug report
[ ] feature request
[ ] support request => Please do not submit support requests here, use one of these channels: https://forum.ionicframework.com/ or https://stencil-worldwide.slack.com

**Current behavior:**
Error when trying to use the PWA Elements camera with vite:

**Expected behavior:**
Module should load and a modal camera should appear.

**Steps to reproduce:**
Build, run, and push the `SNAP` button.

**Related code:**

https://github.com/DinosaurDad/capacitor-pwa-camera-vite

**Other information:**

Error and call stack:

```
TypeError: Failed to fetch dynamically imported module:
   http://localhost:3000/node_modules/.vite/pwa-action-sheet.entry.js
Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'isProxied')
    at core-f86805ad.js:1176:25
    at step (core-f86805ad.js:45:23)
    at Object.next (core-f86805ad.js:26:53)
    at fulfilled (core-f86805ad.js:17:58)
```

**Comment by Vshow101** [comment]
the project is dead

**Comment by ottojaa** [comment]
Would be nice to get some news / workarounds on this

**Comment by SanderGo** [comment]
I'm getting the same error on a brand new Laravel install. They use Vite now by default instead of Webpack (through Mix). 

Capacitor / Camera / PWA Elements worked fine when building with Webpack. After moving to clean Laravel repo with Vite, I got the 'Failed to fetch dynamically imported module'.

Finally got it solved by using the CDN script, which is oke for now. Hope we can just use Vite to make a nice working build sometime soon.

P.S. In the docs section for 'Including through script tag', the src links are incorrect (double paste):
https://capacitorjs.com/docs/web/pwa-elements

**Comment by AbdullahKhan780** [comment]
I'm also getting the same error! Did anyone get any solution?

**Comment by Tommertom** [comment]
Workaround in index.html

https://github.com/Tommertom/svelte-ionic-mystarter-demo/blob/main/index.html

**Comment by Jeziel-Guiev** [comment]
That yes, with CDN it works well and I was testing details in that way, is that when I need to do it offline the CND does not give the first time but as a temporary solution it is not bad at all.

**Comment by wmadden** [comment]
It'd be great to hear from the pwa-elements team on this one, it's been open for a while

**Comment by o-alexandrov** [comment]
It's actually simple, add the following to your `vite.config.ts`, see optimizeDeps in vite's docs.

```
optimizeDeps: {
  exclude: [`@ionic/pwa-elements/loader`],
},
```

Then, in your app, make sure you import as the docs say:
```ts
import { defineCustomElements } from "@ionic/pwa-elements/loader"
```

**Comment by danielroe** [comment]
Is there a way we can make this easier within this module, perhaps? Likely we can add the exclude by default.

**Comment by suchorski** [comment]
> It's actually simple, add the following to your `vite.config.ts`, see [optimizeDeps](https://vitejs.dev/config/dep-optimization-options.html) in `vite`'s docs.
> 
> 
> 
> ```
> 
> optimizeDeps: {
> 
>   exclude: [`@ionic/pwa-elements/loader`],
> 
> },
> 
> ```
> 
> 
> 
> Then, in your app, make sure you import as the docs say:
> 
> ```ts
> 
> import { defineCustomElements } from "@ionic/pwa-elements/loader"
> 
> ```
> 
> 

It works. But when using action sheet, is missing pwa-action-sheet.entry.js when building the dist folder. But when using dev mode, it loads normally from node_modules folder.

Am I missing something? Thanks

**Comment by b4-io** [comment]
same thing happening to me with pwa-camera-modal.entry.js for de camera

**Comment by suchorski** [comment]
> same thing happening to me with pwa-camera-modal.entry.js for de camera
> 
> 

Any news about it? Are you using nuxt?

**Comment by b4-io** [comment]
> > same thing happening to me with pwa-camera-modal.entry.js for de camera
> 
> Any news about it? Are you using nuxt?

Im working on patching the package so it doesn't use any dynamic imports, will be slower on initial render but will work. 

I'm using solidjs.

**Comment by b4-io** [comment]
Patching the package with (patch-package)[https://www.npmjs.com/package/patch-package] seemed to work, it may not be elegant but it is what it is.

I changed the loadModule function in core-... .js. This was my change for the camera only (as fixing the other will result in larger bundles)

```js
import {pwa_camera_modal} from './pwa-camera-modal.entry.js';
import {pwa_camera_modal_instance} from './pwa-camera-modal-instance.entry.js';
import {pwa_camera} from './pwa-camera.entry.js';

var loadModule = function (cmpMeta, hostRef, hmrVersionId) {
    // loadModuleImport
    var exportName = cmpMeta.$tagName$.replace(/-/g, '_');
    var bundleId = (cmpMeta.$lazyBundleIds$);
    var module = moduleCache.get(bundleId);
    if (module) {
        return module[exportName];
    }

    switch (bundleId) {
        case 'pwa-camera-modal': return pwa_camera_modal;
        case 'pwa-camera-modal-instance': return pwa_camera_modal_instance;
        case 'pwa-camera': return pwa_camera;

        default: return import(
            /* webpackInclude: /\.entry\.js$/ */
            /* webpackExclude: /\.system\.entry\.js$/ */
            /* webpackMode: "lazy" */
            "./" + bundleId + ".entry.js" + '').then(function (importedModule) {
                {
                    moduleCache.set(bundleId, importedModule);
                }
                return importedModule[exportName];
            }, consoleError);
    }

};
```

**Comment by sultanmyrza** [comment]
ionic 7 is released 🎉. Will this be fixed by @ionic-team?

**Comment by Erudition** [comment]
> @suchorski Ended up with this function that works with dynamic imports and with the same logic as the original one. Also undid the change to vite config as this fix development too.
> 
> ```js
> var loadModule = function (cmpMeta, hostRef, hmrVersionId) {
>     // loadModuleImport
>     var exportName = cmpMeta.$tagName$.replace(/-/g, '_');
>     var bundleId = (cmpMeta.$lazyBundleIds$);
>     var module = moduleCache.get(bundleId);
>     if (module) {
>         return module[exportName];
>     }
> 
>     const postImport = (importedModule) => {
>         {
>             moduleCache.set(bundleId, importedModule);
>         }
>         return importedModule[exportName];
>     };
> 
>     switch (bundleId) {
>         case 'pwa-camera-modal': return import('./pwa-camera-modal.entry.js').then(postImport, consoleError)
>         case 'pwa-camera-modal-instance': return import('./pwa-camera-modal-instance.entry.js').then(postImport, consoleError)
>         case 'pwa-camera': return import('./pwa-camera.entry.js').then(postImport, consoleError)
>         case 'pwa-action-sheet': return import('./pwa-action-sheet.entry.js').then(postImport, consoleError)
>         case 'pwa-toast': return import('./pwa-toast.entry.js').then(postImport, consoleError)
>     }
> };
> ```

Where is core-**.js?

**Comment by suchorski** [comment]
> WORKS!!

Are you using nuxt 3?

**Comment by Erudition** [comment]
Nope, no framework.

**Comment by suchorski** [comment]
Thanks

**Comment by viniciusgati** [comment]
i dont know where to put the fix... i have no core*js file in my ionic project

**Comment by b4-io** [comment]
> i dont know where to put the fix... i have no core*js file in my ionic project

It's inside `node_modules`

**Comment by sean-perkins** [comment]
Hello everyone 👋 this should be resolved in v3.2.0 of `@ionic/pwa-elements`. 

Let me know if that isn't the case in a new bug report 👍

**Comment by viniciusgati** [comment]
> Hello everyone 👋 this should be resolved in v3.2.0 of `@ionic/pwa-elements`.
> 
> Let me know if that isn't the case in a new bug report 👍

Solved here, thanks man.
